New `Yoast.Commenting.FileComment` sniff

jrfnl opened the pull request:
This sniff is a wrapper around the `Squiz.Commenting.FileComment` sniff as included in the `WordPress-Docs` ruleset.

The wrapper checks whether the file is namespaced or not.

* If the file is namespaced (whether scoped or unscoped) and has a file comment, throw a warning about the file comment being unnecessary.
* If the file is namespaced, but without a namespace name, the normal rules from the `Squiz` sniff apply.

Includes unit test testing the logic in the wrapper. The logic in the Squiz sniff is tested upstream.

Fixes #106

### Remove redundant file comments

Remove file comments from the existing PHP files in this library to comply with the new `Yoast.Commenting.FileComment` sniff.

jrfnl committed:
:sparkles: New `Yoast.Commenting.FileComment` sniff

This sniff is a wrapper around the `Squiz.Commenting.FileComment` sniff as included in the `WordPress-Docs` ruleset.

The wrapper checks whether the the file is namespaced or not.

* If the file is namespaced (whether scoped or unscoped) and has a file comment, throw a warning about the file comment being unnecessary.
* If the file is namespaced, but without a namespace name, the normal rules from the `Squiz` sniff apply.

Includes unit test testing the logic in the wrapper. The logic in the Squiz sniff is tested upstream.

Fixes 106

jrfnl committed:
FileComment: Improve the documentation about the parentage of the sniff

jrfnl committed:
Remove redundant file comments

Remove file comments from the existing PHP files in this library to comply with the new `Yoast.Commenting.FileComment` sniff.

moorscode commented on Yoast/Sniffs/Commenting/FileCommentSniff.php:
```diff
@@ -0,0 +1,65 @@
+			|| $tokens[ $next_non_empty ]['code'] === T_NS_SEPARATOR
+		) {
+			// Either live coding, global namespace (i.e. not really namespaced) or namespace operator.
+			// Normal rules apply.
```
"Normal rules" are the Squiz rules, or parent rules.
Using "normal" requires some mental work to figure out what functionality will be triggered.

moorscode commented on Yoast/Sniffs/Commenting/FileCommentSniff.php:
```diff
@@ -0,0 +1,65 @@
+
+		$namespace_token = $phpcsFile->findNext( T_NAMESPACE, $stackPtr );
+		if ( $namespace_token === false ) {
+			// No namespace found, normal rules apply.
```
"Normal rules" are the Squiz rules, or parent rules.
Using "normal" requires some mental work to figure out what functionality will be triggered.

moorscode commented on Yoast/Sniffs/Commenting/FileCommentSniff.php:
```diff
@@ -0,0 +1,65 @@
+			return parent::process( $phpcsFile, $stackPtr );
+		}
+
+
```
Abundant new-line detected.

moorscode commented on Yoast/Sniffs/Commenting/FileCommentSniff.php:
```diff
@@ -0,0 +1,65 @@
+
+		if ( $tokens[ $comment_start ]['code'] === T_DOC_COMMENT_OPEN_TAG ) {
+			$phpcsFile->addWarning(
+				'A namespaced file does not need a file docblock',
```
`A file containing namespaced code, does not need a file docblock.`
"Namespaced file" could also refer to the filename or directory structure.

jrfnl committed:
FileComment: Fix the tests

As the unit tests are run without the context of the `ruleset.xml` file, the exclusions made in that file will not be applied.

For that reason, the unit test files which test that the sniff correctly falls through to the Squiz parent sniff, need to comply with the full rules from the `Squiz.Commenting.FileComment` sniff, including the Squiz author/copyright tags.

jrfnl commented:
@moorscode Thanks for the review. Good points.

I've added two additional commits: One to address the comments above. The second to fix the unit tests.

> As the unit tests are run without the context of the `ruleset.xml` file, the exclusions made in that file will not be applied.
> 
> For that reason, the unit test files which test that the sniff correctly falls through to the Squiz parent sniff, need to comply with the full rules from the `Squiz.Commenting.FileComment` sniff, including the Squiz author/copyright tags.

(from the commit msg)

I've added these changes in separate commits for easier review, but as this is a new sniff, I'd be happy to squash these last two commits into the very first one once the PR is approved (but not merged yet).

atimmer commented:
👍 , sniff and test cases look good.

moorscode commented:
:+1: all looks good
